spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -324,6 +363,22 @@ function chanUpgradeTry(
     // verify proofs of counterparty state
     abortTransactionUnless(verifyChannelState(connection, proofHeight, proofChannel, currentChannel.counterpartyPortIdentifier, currentChannel.counterpartyChannelIdentifier, counterpartyChannel))
     abortTransactionUnless(verifyChannelUpgradeTimeout(connection, proofHeight, proofUpgradeTimeout, currentChannel.counterpartyPortIdentifier, currentChannel.counterpartyChannelIdentifier, upgradeTimeout))
+    abortTransactionUnless(verifyUpgradeSequence(connection, proofHeight, proofUpgradeSequence, currentChannel.counterpartyPortIdentifier,
+    currentChannel.counterpartyChannelIdentifier, counterpartySequence))
+
+    // get current sequence on this channel
+    // if the counterparty sequence is greater than the current sequence, we fast forward to the counterparty sequence
+    // so that both channel ends are using the same sequence for the current upgrade
+    // if the counterparty sequence is less than the current sequence, then either the counterparty chain is out-of-sync or
+    // the message is out-of-sync and we write an error receipt with our own sequence so that the counterparty can update
+    // their sequence as well.
+    currentSequence = provableStore.get(upgradeSequencePath(portIdentifier, channelIdentifier))
+    if counterpartySequence >= currentSequence {
+        provableStore.set(upgradeSequencePath(portIdentifier, channelIdentifier), counterpartySequence)
+    } else {
+        provableStore.set(errorPath(portIdentifier, channelIdentifier, currentSequence), []byte{1})
```

Comment:
I don't believe this error condition will be hit, but please review that this is a correct recovery in case the sequences do go out of sync

Reply:
I believe that the error condition is hit in the following scenario involving two chains A and B. Both have sequence `10.` 
- A executes INIT and sends `ChanUpgradeTry(seq=10)`
- B executes TRY, however it results in restore, which entails `errorB = 10` and `seqB=11`
- A does not receive `ChanUpgradeCancelMsg` (`seqA` remains set to `10`)
- A eventually receives `ChanUpgradeTimeoutMsg`, but `seqA` still remains set to `10`
- A executes INIT and sends `ChanUpgradeTry(seq=10)`
- B executes TRY and writes `errorB = 11`

Moreover, 
- A receives `ChanUpgradeCancelMsg(seq=11)`, sets `seqA = 11` and aborts the transaction
- A executes INIT and sends `ChanUpgradeTry(seq=11)`
- B executes TRY and sends `ChanUpgradeAck()`
- a relayer relays `ChanUpgradeCancelMsg(seq=11)` to A
  - A aborts the transaction and sets `seqA = 12`
  - B gets stuck in `UPGRADE_TRY`

Reply:
A solution would be to have the following changes:
```typescript
function chanUpgradeTry {
  // .... 
  currentSequence = provableStore.get(upgradeSequencePath(portIdentifier, channelIdentifier))
  if counterpartySequence >= currentSequence {
    provableStore.set(upgradeSequencePath(portIdentifier, channelIdentifier), counterpartySequence)
  } else {
    provableStore.set(errorPath(portIdentifier, channelIdentifier, currentSequence), []byte{1})
    // CHANGE: increment sequence similarly to restore()
    return
  }
  // .... 
}
```

```typescript
function cancelChannelUpgrade {
  // .... 
  currentSequence = provableStore.get(currentSequencePath(portIdentifier, channelIdentifier))
  abortTransactionUnless(counterpartySequence >= currentSequence)
  if counterpartySequence > currentSequence {
    // the counterparty sequence is higher and thus the channel ends are out of sync
    // setting our sequence to the counterparty sequence will bring the channel ends back in sync
    // CHANGE: additionally increment the sequence, i.e., counterpartySequence+1
    provableStore.set(upgradeSequencePath(portIdentifier, channelIdentifier), counterpartySequence+1)
  } else {
    // current sequence and counterparty sequence are equal so both chains are in sync.
    // We must increment the sequence to prepare for the next upgrade
    provableStore.set(upgradeSequencePath(portIdentifier, channelIdentifier), currentSequence+1)
  }
  // .... 
}
```
Basically, if chain A attempts an upgrade for a sequence lower than the sequence on chan B, then the sequences on both A and B should not be used again. Not sure though if this has further implications on other parts of the protocol.

Reply:
> Moreover,
> A receives ChanUpgradeCancelMsg(seq=11), sets seqA = 11 and aborts the transaction
> A executes INIT and sends ChanUpgradeTry(seq=11)
> B executes TRY and sends ChanUpgradeAck()
> a relayer relays ChanUpgradeCancelMsg(seq=11) to A
> A aborts the transaction and sets seqA = 12
> B gets stuck in UPGRADE_TRY

This is absolutely correct. We have to bump both sequences on an error. This is simple to do and I'll make the requisite changes.

> A executes INIT and sends ChanUpgradeTry(seq=10)
> B executes TRY, however it results in restore, which entails errorB = 10 and seqB=11
> A does not receive ChanUpgradeCancelMsg (seqA remains set to 10)
> A eventually receives ChanUpgradeTimeoutMsg, but seqA still remains set to 10
> A executes INIT and sends ChanUpgradeTry(seq=10)
> B executes TRY and writes errorB = 11

This can never happen precisely because Timeouts can only be processed by proving that the counterparty had not seen the TRY message (same as packet timeouts, also see timeout pseudocode). So regardless of how much time passes after B restores connection, the next step in the handshake must be a `ChanUpgradeCancelMsg` not a `ChanUpgradeTimeout` which would fail. The channel will simply be unlive during that time.